From 61d11391080d0e5665250729b4bdf18a79a9c1ec Mon Sep 17 00:00:00 2001
From: Kevin Carter <kevin.carter@rackspace.com>
Date: Wed, 11 Nov 2015 10:32:03 -0600
Subject: [PATCH] Functional back port of the Liberty repo-build process

The change back ports the repo-build role into kilo such that
we're allowing the Kilo build process to benefit from the changes
made in Liberty+. This will also allow for the Kilo playbooks
to clone from a repo using a generated manifest file. This will
ensure that the build process is more repeatable, creating an
environment with faster build times and drop the dependency
on the yaprt library.

This change brings with it all of the associated updates that have been
made to the various plugins and playbooks to ensure a better repo-env.

* the repo-* plays have been back ported from Liberty
* the repo-build role has been back ported
* the py_pkg lookup plugin changes have been back ported

Change-Id: I5a3a1506bf3df8b5812d9131248d0156a8dd7dd6
Signed-off-by: Kevin Carter <kevin.carter@rackspace.com>
---

diff --git a/playbooks/defaults/repo_packages/openstack_other.yml b/playbooks/defaults/repo_packages/openstack_other.yml
index 4294614..85d0c0f 100644
--- a/playbooks/defaults/repo_packages/openstack_other.yml
+++ b/playbooks/defaults/repo_packages/openstack_other.yml
@@ -28,7 +28,7 @@
 ## Tempest service
 tempest_git_repo: https://git.openstack.org/openstack/tempest
 tempest_git_install_branch: d22ec33640affaeabd17c591599da86ffdf6d722 # HEAD of "master" as of 21.10.2015
-tempest_git_install_fragments: "yaprtignorerequirements=true"
+tempest_git_install_fragments: "ignorerequirements=true"
 tempest_git_dest: "/opt/tempest_{{ tempest_git_install_branch | replace('/', '_') }}"
 
 # NOVNC from source
@@ -40,3 +40,8 @@
 spicehtml5_git_repo: https://github.com/SPICE/spice-html5
 spicehtml5_git_install_branch: c1e736b083ff47639ecb73ea9be4d14b5002f93f # HEAD of "master" as of 21.10.2015
 spicehtml5_git_dest: "/opt/spicehtml5_{{ spicehtml5_git_install_branch | replace('/', '_') }}"
+
+# Pinned because any newer version will break within Kilo
+openstackclient_git_repo: https://git.openstack.org/openstack/python-openstackclient
+openstackclient_git_install_branch: e750e5784dcb4b58ef2359f9dded9064a92b58a3 # HEAD of "tag/1.0.4" as of 16.04.2015
+openstackclient_git_dest: "/opt/openstackclient_{{ openstackclient_git_install_branch | replace('/', '_') }}"
diff --git a/playbooks/defaults/repo_packages/openstack_services.yml b/playbooks/defaults/repo_packages/openstack_services.yml
index 1fce9c1..118ba81 100644
--- a/playbooks/defaults/repo_packages/openstack_services.yml
+++ b/playbooks/defaults/repo_packages/openstack_services.yml
@@ -55,7 +55,7 @@
 
 ## Heat service
 heat_git_repo: https://git.openstack.org/openstack/heat
-heat_git_install_branch: a4aafba0c2b387df888dfd805594b78b39b5f495 # HEAD of "stable/kilo" as of 21.10.2015
+heat_git_install_branch: 81de0a272cf7973047571bf7677052025647c192 # HEAD of "stable/kilo" as of 11.11.2015
 heat_git_dest: "/opt/heat_{{ heat_git_install_branch | replace('/', '_') }}"
 heat_repo_plugins:
   - { path: "contrib", package: "extraroute" }
diff --git a/playbooks/plugins/filters/osa-filters.py b/playbooks/plugins/filters/osa-filters.py
index e532d14..61ca112 100644
--- a/playbooks/plugins/filters/osa-filters.py
+++ b/playbooks/plugins/filters/osa-filters.py
@@ -14,6 +14,13 @@
 #
 # (c) 2015, Kevin Carter <kevin.carter@rackspace.com>
 
+import os
+import re
+import urlparse
+import hashlib
+
+from ansible import errors
+
 """Filter usage:
 
 Simple filters that may be useful from within the stack
@@ -30,11 +37,170 @@
     return 2**(int(value)-1).bit_length()
 
 
+def get_netloc(url):
+    """Return the netloc from a URL.
+
+    If the input value is not a value URL the method will raise an Ansible
+    filter exception.
+
+    :param url: the URL to parse
+    :type url: ``str``
+    :returns: ``str``
+    """
+    try:
+        netloc = urlparse.urlparse(url).netloc
+    except Exception as exp:
+        raise errors.AnsibleFilterError(
+            'Failed to return the netloc of: "%s"' % str(exp)
+        )
+    else:
+        return netloc
+
+
+def get_netloc_no_port(url):
+    """Return the netloc without a port from a URL.
+
+    If the input value is not a value URL the method will raise an Ansible
+    filter exception.
+
+    :param url: the URL to parse
+    :type url: ``str``
+    :returns: ``str``
+    """
+    return get_netloc(url=url).split(':')[0]
+
+
+def get_netorigin(url):
+    """Return the netloc from a URL.
+
+    If the input value is not a value URL the method will raise an Ansible
+    filter exception.
+
+    :param url: the URL to parse
+    :type url: ``str``
+    :returns: ``str``
+    """
+    try:
+        parsed_url = urlparse.urlparse(url)
+        netloc = parsed_url.netloc
+        scheme = parsed_url.scheme
+    except Exception as exp:
+        raise errors.AnsibleFilterError(
+            'Failed to return the netorigin of: "%s"' % str(exp)
+        )
+    else:
+        return '%s://%s' % (scheme, netloc)
+
+
+def string_2_int(string):
+    """Return the an integer from a string.
+
+    The string is hashed, converted to a base36 int, and the modulo of 10240
+    is returned.
+
+    :param string: string to retrieve an int from
+    :type string: ``str``
+    :returns: ``int``
+    """
+    # Try to encode utf-8 else pass
+    try:
+        string = string.encode('utf-8')
+    except AttributeError:
+        pass
+    hashed_name = hashlib.sha256(string).hexdigest()
+    return int(hashed_name, 36) % 10240
+
+
+def pip_requirement_names(requirements):
+    """Return a ``str`` of requirement name and list of versions.
+    :param requirement: Name of a requirement that may have versions within
+                        it. This will use the constant,
+                        VERSION_DESCRIPTORS.
+    :type requirement: ``str``
+    :return: ``str``
+    """
+
+    version_descriptors = "(>=|<=|>|<|==|~=|!=)"
+    named_requirements = list()
+    for requirement in requirements:
+        requirement = requirement.split(';')[0]
+        name = re.split(r'%s\s*' % version_descriptors, requirement)[0]
+        if name and not name.startswith('#'):
+            named_requirements.append(name.lower())
+
+    return sorted(set(named_requirements))
+
+
+def splitlines(string_with_lines):
+    """Return a ``list`` from a string with lines."""
+
+    return string_with_lines.splitlines()
+
+
+def filtered_list(list_one, list_two):
+
+    _list_one = set([i.lower() for i in list_one])
+    _list_two = set([i.lower() for i in list_two])
+    return list(_list_one-_list_two)
+
+
+def git_link_parse(repo):
+    """Return a dict containing the parts of a git repository.
+
+    :param repo: git repo string to parse.
+    :type repo: ``str``
+    :returns: ``dict``
+    """
+
+    if 'git+' in repo:
+        _git_url = repo.split('git+', 1)[-1]
+    else:
+        _git_url = repo
+
+    if '@' in _git_url:
+        url, branch = _git_url.split('@', 1)
+    else:
+        url = _git_url
+        branch = 'master'
+
+    name = os.path.basename(url.rstrip('/'))
+    _branch = branch.split('#')
+    branch = _branch[0]
+
+    plugin_path = None
+    # Determine if the package is a plugin type
+    if len(_branch) > 1 and 'subdirectory=' in _branch[-1]:
+        plugin_path = _branch[-1].split('subdirectory=')[-1].split('&')[0]
+
+    return {
+        'name': name.split('.git')[0].lower(),
+        'version': branch,
+        'plugin_path': plugin_path,
+        'url': url,
+        'original': repo
+    }
+
+
+def git_link_parse_name(repo):
+    """Return the name of a git repo."""
+
+    return git_link_parse(repo)['name']
+
+
 class FilterModule(object):
     """Ansible jinja2 filters."""
 
     @staticmethod
     def filters():
         return {
-            'bit_length_power_of_2': bit_length_power_of_2
+            'bit_length_power_of_2': bit_length_power_of_2,
+            'netloc': get_netloc,
+            'netloc_no_port': get_netloc_no_port,
+            'netorigin': get_netorigin,
+            'string_2_int': string_2_int,
+            'pip_requirement_names': pip_requirement_names,
+            'splitlines': splitlines,
+            'filtered_list': filtered_list,
+            'git_link_parse': git_link_parse,
+            'git_link_parse_name': git_link_parse_name
         }
diff --git a/playbooks/plugins/lookups/py_pkgs.py b/playbooks/plugins/lookups/py_pkgs.py
index ac1ffb6..004ad31 100644
--- a/playbooks/plugins/lookups/py_pkgs.py
+++ b/playbooks/plugins/lookups/py_pkgs.py
@@ -26,10 +26,10 @@
 
 
 REQUIREMENTS_FILE_TYPES = [
-    'requirements.txt',
     'global-requirements.txt',
     'test-requirements.txt',
-    'dev-requirements.txt'
+    'dev-requirements.txt',
+    'global-requirement-pins.txt'
 ]
 
 
@@ -43,6 +43,61 @@
 ]
 
 
+def git_pip_link_parse(repo):
+    """Return a tuple containing the parts of a git repository.
+
+    Example parsing a standard git repo:
+      >>> git_pip_link_parse('git+https://github.com/username/repo@tag')
+      ('repo',
+       'tag',
+       None,
+       'https://github.com/username/repo',
+       'git+https://github.com/username/repo@tag')
+
+    Example parsing a git repo that uses an installable from a subdirectory:
+      >>> git_pip_link_parse(
+      ...     'git+https://github.com/username/repo@tag#egg=plugin.name'
+      ...     '&subdirectory=remote_path/plugin.name'
+      ... )
+      ('repo',
+       'tag',
+       'remote_path/plugin.name',
+       'https://github.com/username/repo',
+       'git+https://github.com/username/repo@tag#egg=plugin.name&'
+       'subdirectory=remote_path/plugin.name')
+
+    :param repo: git repo string to parse.
+    :type repo: ``str``
+    :returns: ``tuple``
+    """
+
+    _git_url = repo.split('+')
+    if len(_git_url) >= 2:
+        _git_url = _git_url[1]
+    else:
+        _git_url = _git_url[0]
+
+    git_branch_sha = _git_url.split('@')
+    if len(git_branch_sha) > 1:
+        url, branch = git_branch_sha
+    else:
+        url = git_branch_sha[0]
+        branch = 'master'
+
+    name = os.path.basename(url.rstrip('/'))
+    _branch = branch.split('#')
+    branch = _branch[0]
+
+    plugin_path = None
+    # Determine if the package is a plugin type
+    if len(_branch) > 1:
+        if 'subdirectory' in _branch[-1]:
+            plugin_path = _branch[1].split('subdirectory=')[1].split('&')[0]
+            name = os.path.basename(plugin_path)
+
+    return name.lower(), branch, plugin_path, url, repo
+
+
 class DependencyFileProcessor(object):
     def __init__(self, local_path):
         """Find required files.
@@ -53,6 +108,7 @@
         self.pip = dict()
         self.pip['git_package'] = list()
         self.pip['py_package'] = list()
+        self.pip['role_packages'] = dict()
         self.git_pip_install = 'git+%s@%s'
         self.file_names = self._get_files(path=local_path)
 
@@ -67,9 +123,10 @@
         :returns: ``list``
         """
         _file_names = list()
+        file_name_words = ['/defaults/', '/vars/', '/user_']
         for file_name in file_names:
             if file_name.endswith(ext):
-                if '/defaults/' in file_name or '/vars/' in file_name:
+                if any(i in file_name for i in file_name_words):
                     _file_names.append(file_name)
                 else:
                     continue
@@ -122,6 +179,7 @@
 
             if git_data['fragments']:
                 package = '%s&%s' % (package, git_data['fragments'])
+                name = os.path.basename(plugin_path)
 
             self.pip['git_package'].append(package)
 
@@ -151,7 +209,9 @@
         package = self.git_pip_install % (
             git_data['repo'], git_data['branch']
         )
-
+        package = '%s#egg=%s' % (
+            package, git_pip_link_parse(package)[0].replace('-', '_')
+        )
         git_data['fragments'] = loaded_yaml.get(
             '%s_git_install_fragments' % var_name.replace('.', '_')
         )
@@ -177,6 +237,7 @@
             ext=ext
         )
 
+        role_name = None
         for file_name in file_names:
             with open(file_name, 'rb') as f:
                 # If there is an exception loading the file continue
@@ -190,15 +251,35 @@
                     if not loaded_config:
                         continue
 
+                    if 'roles' in file_name:
+                        _role_name = file_name.split('roles%s' % os.sep)[-1]
+                        role_name = _role_name.split(os.sep)[0]
+
+
             for key, values in loaded_config.items():
-                if key.endswith('git_repo'):
-                    self._process_git(
-                        loaded_yaml=loaded_config,
-                        git_item=key
-                    )
+                # This conditional is set to ensure we're not processes git repos
+                #  from the defaults file which may conflict with what is being set
+                #  in the repo_packages files.
+                if not '/defaults/main' in file_name:
+                    if key.endswith('git_repo'):
+                        self._process_git(
+                            loaded_yaml=loaded_config,
+                            git_item=key
+                        )
 
                 if [i for i in BUILT_IN_PIP_PACKAGE_VARS if i in key]:
                     self.pip['py_package'].extend(values)
+
+                    if role_name:
+                        if not role_name in self.pip['role_packages']:
+                            self.pip['role_packages'][role_name] = values
+                        else:
+                            self.pip['role_packages'][role_name].extend(values)
+                            self.pip['role_packages'][role_name] = sorted(
+                                set(
+                                    self.pip['role_packages'][role_name]
+                                )
+                            )
 
 
 def _abs_path(path):
@@ -231,8 +312,14 @@
         if isinstance(terms, basestring):
             terms = [terms]
 
-        return_list = list()
+        return_data = {
+            'packages': set(),
+            'remote_packages': set(),
+            'remote_package_parts': list(),
+            'role_packages': dict()
+        }
         for term in terms:
+            return_list = list()
             try:
                 dfp = DependencyFileProcessor(
                     local_path=_abs_path(str(term))
@@ -247,27 +334,55 @@
                         traceback.format_exc()
                     )
                 )
-        else:
-            return_data = {
-                'packages': list(),
-                'remote_packages': list()
-            }
-            for file_name in sorted(set(return_list)):
-                is_url = file_name.startswith(('http:', 'https:', 'git+'))
-                if is_url:
-                    if '@' not in file_name:
-                        return_data['packages'].append(file_name)
+
+            for item in sorted(set(return_list)):
+                if item.startswith(('http:', 'https:', 'git+')):
+                    if '@' not in item:
+                        return_data['packages'].add(item)
                     else:
-                        return_data['remote_packages'].append(file_name)
+                        item_name = git_pip_link_parse(item)[0]
+                        # loaded as a list to avoid a RuntimeError:
+                        #  Set changed size during iteration
+                        for rpkg in list(return_data['remote_packages']):
+                            rpkg_name = git_pip_link_parse(rpkg)[0]
+                            if rpkg_name == item_name:
+                                return_data['remote_packages'].remove(rpkg)
+                        else:
+                            return_data['remote_packages'].add(item)
                 else:
-                    return_data['packages'].append(file_name)
+                    return_data['packages'].add(item)
             else:
-                return_data['packages'] = ' '.join(
-                    ['"%s"' % i for i in set(return_data['packages'])]
-                )
+                return_data['packages'] = [
+                    i.lower() for i in return_data['packages']
+                ]
 
-                return_data['remote_packages'] = ' '.join(
-                    ['"%s"' % i for i in set(return_data['remote_packages'])]
+                keys = ['name', 'version', 'fragment', 'url', 'original']
+                remote_pkg_parts = [
+                    dict(
+                        zip(
+                            keys, git_pip_link_parse(i)
+                        )
+                    ) for i in return_data['remote_packages']
+                ]
+                return_data['remote_package_parts'].extend(remote_pkg_parts)
+                return_data['remote_package_parts'] = list(
+                    dict(
+                        (i['name'], i)
+                        for i in return_data['remote_package_parts']
+                    ).values()
                 )
+                return_data['role_packages'].update(dfp.pip['role_packages'])
+        else:
+            # Sort everything within the returned data
+            for key, value in return_data.items():
+                if isinstance(value, (list, set)):
+                    return_data[key] = sorted(value)
+            return [return_data]
 
-                return [return_data]
+
+# Used for testing and debuging usage: `python plugins/lookups/py_pkgs.py ../`
+if __name__ == '__main__':
+    import sys
+    import json
+    print(json.dumps(LookupModule().run(terms=sys.argv[1:]), indent=4))
+
diff --git a/playbooks/repo-build.yml b/playbooks/repo-build.yml
index 0858b09..e11019e 100644
--- a/playbooks/repo-build.yml
+++ b/playbooks/repo-build.yml
@@ -1,5 +1,5 @@
 ---
-# Copyright 2014, Rackspace US, Inc.
+# Copyright 2015, Rackspace US, Inc.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
@@ -13,91 +13,31 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# The purpose here is to allow for the environment to update/build the
-# python wheel files from the CURRENT release, as set in the openstack_release
-# variable.
 - name: Build new repo packages for a given release
   hosts: repo_all[0]
-  max_fail_percentage: 20
-  gather_facts: false
+  gather_facts: true
   user: root
-  tasks:
-    - name: Create a build report for all known packages within a release
-      shell: |
-        yaprt --quiet \
-              create-report \
-              --report-file {{ repo_service_home_folder }}/repo/reports/release-{{ openstack_release }}-report.json \
-              --git-install-repos {{ item['remote_packages'] }} \
-              --git-repo-path "{{ repo_service_home_folder }}/repo/openstackgit" \
-              --packages {{ item['packages'] }}
+  pre_tasks:
+    - name: Load local packages
+      debug:
+        msg: "Loading Packages"
       with_py_pkgs: ../
-      sudo: yes
-      sudo_user: "{{ repo_service_user_name }}"
+      register: local_packages
       tags:
-        - repo-create-report
-
-    - name: Build all known python packages requirements
-      shell: |
-        yaprt --quiet \
-              build-wheels \
-              --report-file {{ repo_service_home_folder }}/repo/reports/release-{{ openstack_release }}-report.json \
-              --storage-pool "{{ repo_service_home_folder }}/repo/pools" \
-              --link-dir "{{ repo_service_home_folder }}/repo/os-releases/{{ openstack_release }}" \
-              --pip-extra-link-dirs "{{ repo_service_home_folder }}/repo/links" \
-              --pip-index "{{ repo_pip_default_index }}" \
-              --pip-extra-index "https://pypi.python.org/simple/" \
-              --pip-bulk-operation \
-              --build-output "{{ repo_build_output }}" \
-              --build-dir "{{ repo_build_dir }}" \
-              --build-requirements \
-              --git-repo-path "{{ repo_service_home_folder }}/repo/openstackgit" \
-              --force-clean
-      sudo: yes
-      sudo_user: "{{ repo_service_user_name }}"
+        - repo-clone-repos
+        - repo-set-requirements
+        - repo-get-global-requirements
+        - repo-set-requirement-names
+        - repo-set-requirement-names-filtered
+        - repo-set-constraints
+        - repo-build-constraints-file
+        - repo-create-venv-archive
+        - repo-venv-compress-archive
+        - repo-build-venvs
+  roles:
+    - role: "repo_build"
+      repo_build_release_tag: "{{ openstack_release }}"
+      repo_build_pip_default_index: "https://rpc-repo.rackspace.com/pools"
+      repo_build_pip_extra_index: "https://pypi.python.org/simple"
       tags:
-        - repo-build-requirements
-
-    - name: Build all known python packages git sources
-      shell: |
-        yaprt --quiet \
-              build-wheels \
-              --report-file {{ repo_service_home_folder }}/repo/reports/release-{{ openstack_release }}-report.json \
-              --storage-pool "{{ repo_service_home_folder }}/repo/pools" \
-              --link-dir "{{ repo_service_home_folder }}/repo/os-releases/{{ openstack_release }}" \
-              --pip-extra-link-dirs "{{ repo_service_home_folder }}/repo/links" \
-              --pip-no-deps \
-              --build-output "{{ repo_build_output }}" \
-              --build-dir "{{ repo_build_dir }}" \
-              --build-branches \
-              --build-releases \
-              --git-repo-path "{{ repo_service_home_folder }}/repo/openstackgit" \
-              --force-clean
-      sudo: yes
-      sudo_user: "{{ repo_service_user_name }}"
-      tags:
-        - repo-build-git-sources
-
-    - name: Create html indexes
-      shell: |
-        yaprt --quiet \
-              create-html-indexes \
-              --repo-dir "{{ repo_service_home_folder }}/repo" \
-              --dir-exclude "{{ repo_service_home_folder }}/repo/openstackgit"
-      sudo: yes
-      sudo_user: "{{ repo_service_user_name }}"
-      tags:
-        - repo-html-indexes
-
-    - name: Store all git source
-      shell: |
-        yaprt --quiet \
-              store-repos \
-              --report-file {{ repo_service_home_folder }}/repo/reports/release-{{ openstack_release }}-report.json \
-              --git-repo-path "{{ repo_service_home_folder }}/repo/openstackgit"
-      sudo: yes
-      sudo_user: "{{ repo_service_user_name }}"
-      tags:
-        - repo-store-git-sources
-  vars:
-    repo_build_dir: "/tmp/openstack-builder"
-    repo_build_output: "/tmp/openstack-wheel-output"
+        - "repo-build"
diff --git a/playbooks/repo-clone-mirror.yml b/playbooks/repo-clone-mirror.yml
index b9303f8..b08deb7 100644
--- a/playbooks/repo-clone-mirror.yml
+++ b/playbooks/repo-clone-mirror.yml
@@ -1,5 +1,5 @@
 ---
-# Copyright 2014, Rackspace US, Inc.
+# Copyright 2015, Rackspace US, Inc.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
@@ -17,27 +17,52 @@
 # from within the environment as found from a given mirror_source_host.
 # Currently the mirror source host is set to the Rackspace build servers but
 # could be targeted to wherever you'd like.
+
 - name: Cloning the upstream repo mirror
   hosts: repo_all[0]
-  max_fail_percentage: 20
   gather_facts: false
   user: root
+  pre_tasks:
+    - name: Delete old MANIFEST file if found
+      file:
+        state: absent
+        path: /tmp/MANIFEST.in
   tasks:
+    - name: download MANIFEST.in
+      get_url:
+        url: "{{ repo_upstream_manifest_url }}"
+        dest: /tmp/MANIFEST.in
     - name: Sync the upstream repo(s)
       shell: |
-        rsync -avzlHAX \
-              {{ mirror_excludes }} \
-              {{ mirror_source_host }}::{{ mirror_name }} {{ mirror_path }}
+        {{ rsync_commands }} \
+              {{ rsync_flags }} \
+              --files-from=/tmp/MANIFEST.in \
+              {{ repo_upstream_url | netloc }}::{{ mirror_name }} {{ mirror_path }}
+      sudo: yes
+      sudo_user: "{{ repo_service_user_name }}"
+    - name: Create sync directories
+      file:
+        state: directory
+        path: "{{ mirror_path }}/{{ item }}"
+        owner: "{{ repo_service_user_name }}"
+      with_items: rsync_dirs
+    - name: Sync supporting directories
+      shell: |
+        {{ rsync_commands }} \
+              {{ rsync_flags }} \
+              {{ repo_upstream_url | netloc }}::{{ mirror_name }}/{{ item }}/ {{ mirror_path }}/{{ item }}
+      with_items: rsync_dirs
       sudo: yes
       sudo_user: "{{ repo_service_user_name }}"
   vars:
-    repo_mirror_excludes:
-      - "/repos"
-      - "/mirror"
-      - "/rpcgit"
-      - "/python_packages"
-    mirror_excludes: "{% for i in repo_mirror_excludes %} --exclude={{ i }} {% endfor %}"
+    rsync_commands: rsync
+    rsync_flags: "-avzlHAX"
+    rsync_dirs:
+      - container_images
+      - downloads
+      - "venvs/{{ openstack_release }}"
     mirror_path: "{{ repo_service_home_folder }}/repo"
-    mirror_name: "{{ repo_mirror_name|default('openstack_mirror') }}"
-    mirror_source_host: "{{ repo_mirror_source_host|default('rpc-repo.rackspace.com') }}"
-    is_metal: "{{ properties.is_metal|default(false) }}"
+    mirror_name: "openstack_mirror"
+    repo_upstream_url: "https://rpc-repo.rackspace.com"
+    repo_upstream_manifest_url: "{{ repo_upstream_url }}/os-releases/{{ openstack_release }}/MANIFEST.in"
+    repo_service_user_name: nginx
diff --git a/playbooks/repo-install.yml b/playbooks/repo-install.yml
index 3a299ac..ea20df7 100644
--- a/playbooks/repo-install.yml
+++ b/playbooks/repo-install.yml
@@ -14,4 +14,4 @@
 # limitations under the License.
 
 - include: repo-server.yml
-- include: repo-clone-mirror.yml
+- include: repo-build.yml
diff --git a/playbooks/repo-server.yml b/playbooks/repo-server.yml
index 721f875..b67d696 100644
--- a/playbooks/repo-server.yml
+++ b/playbooks/repo-server.yml
@@ -58,11 +58,17 @@
   roles:
     - { role: "repo_server", tags: [ "repo-server" ] }
     - role: "rsyslog_client"
-      rsyslog_client_log_rotate_file: repo_log_rotate
+      rsyslog_client_log_rotate_file: repo_nginx_log_rotate
       rsyslog_client_log_dir: "/var/log/nginx"
       rsyslog_client_log_files:
-        - /var/www/repo_builder.log
         - /var/log/rsyncd.log
+      rsyslog_client_config_name: "99-repo-nginx-rsyslog-client.conf"
+      tags:
+        - "repo-nginx-rsyslog-client"
+        - "rsyslog-client"
+    - role: "rsyslog_client"
+      rsyslog_client_log_rotate_file: repo_log_rotate
+      rsyslog_client_log_dir: "/var/log/repo"
       rsyslog_client_config_name: "99-repo-rsyslog-client.conf"
       tags:
         - "repo-rsyslog-client"
diff --git a/playbooks/roles/repo_build/CONTRIBUTING.rst b/playbooks/roles/repo_build/CONTRIBUTING.rst
new file mode 100644
index 0000000..4835816
--- /dev/null
+++ b/playbooks/roles/repo_build/CONTRIBUTING.rst
@@ -0,0 +1,102 @@
+OpenStack repo build
+####################
+:tags: openstack, repo, build, cloud, ansible
+:category: \*nix
+
+contributor guidelines
+^^^^^^^^^^^^^^^^^^^^^^
+
+Filing Bugs
+-----------
+
+Bugs should be filed on Launchpad, not GitHub: "https://bugs.launchpad.net/openstack-ansible"
+
+
+When submitting a bug, or working on a bug, please ensure the following
+criteria are met:
+
+* The description clearly states or describes the original problem or root
+  cause of the problem.
+* Include historical information on how the problem was identified.
+* Any relevant logs are included.
+* The provided information should be totally self-contained. External access to
+  web services/sites should not be needed.
+* Steps to reproduce the problem if possible.
+
+
+Submitting Code
+---------------
+
+Changes to the project should be submitted for review via the Gerrit tool,
+following the workflow documented at:
+"http://docs.openstack.org/infra/manual/developers.html#development-workflow"
+
+Pull requests submitted through GitHub will be ignored and closed without
+regard.
+
+
+Extra
+-----
+
+Tags:
+    If it's a bug that needs fixing in a branch in addition to Master, add a
+    '\<release\>-backport-potential' tag (eg ``juno-backport-potential``).
+    There are predefined tags that will autocomplete.
+
+Status:
+    Please leave this alone, it should be New till someone triages the issue.
+
+Importance:
+    Should only be touched if it is a Blocker/Gating issue. If it is, please
+    set to High, and only use Critical if you have found a bug that can take
+    down whole infrastructures.
+
+
+Style guide
+-----------
+
+When creating tasks and other roles for use in Ansible please create then using
+the YAML dictionary format.
+
+Example YAML dictionary format:
+    .. code-block:: yaml
+
+        - name: The name of the tasks
+          module_name:
+            thing1: "some-stuff"
+            thing2: "some-other-stuff"
+          tags:
+            - some-tag
+            - some-other-tag
+
+
+Example **NOT** in YAML dictionary format:
+    .. code-block:: yaml
+
+        - name: The name of the tasks
+          module_name: thing1="some-stuff" thing2="some-other-stuff"
+          tags:
+            - some-tag
+            - some-other-tag
+
+
+Usage of the ">" and "|" operators should be limited to Ansible conditionals
+and command modules such as the ansible ``shell`` module.
+
+
+Issues
+------
+
+When submitting an issue, or working on an issue please ensure the following
+criteria are met:
+
+* The description clearly states or describes the original problem or root
+  cause of the problem.
+* Include historical information on how the problem was identified.
+* Any relevant logs are included.
+* If the issue is a bug that needs fixing in a branch other than Master, add
+  the ‘backport potential’ tag TO THE ISSUE (not the PR).
+* The provided information should be totally self-contained. External access to
+  web services/sites should not be needed.
+* If the issue is needed for a hotfix release, add the 'expedite' label.
+* Steps to reproduce the problem if possible.
diff --git a/playbooks/roles/repo_build/LICENSE b/playbooks/roles/repo_build/LICENSE
new file mode 100644
index 0000000..e06d208
--- /dev/null
+++ b/playbooks/roles/repo_build/LICENSE
@@ -0,0 +1,202 @@
+Apache License
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
+
+   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
+
+   1. Definitions.
+
+      "License" shall mean the terms and conditions for use, reproduction,
+      and distribution as defined by Sections 1 through 9 of this document.
+
+      "Licensor" shall mean the copyright owner or entity authorized by
+      the copyright owner that is granting the License.
+
+      "Legal Entity" shall mean the union of the acting entity and all
+      other entities that control, are controlled by, or are under common
+      control with that entity. For the purposes of this definition,
+      "control" means (i) the power, direct or indirect, to cause the
+      direction or management of such entity, whether by contract or
+      otherwise, or (ii) ownership of fifty percent (50%) or more of the
+      outstanding shares, or (iii) beneficial ownership of such entity.
+
+      "You" (or "Your") shall mean an individual or Legal Entity
+      exercising permissions granted by this License.
+
+      "Source" form shall mean the preferred form for making modifications,
+      including but not limited to software source code, documentation
+      source, and configuration files.
+
+      "Object" form shall mean any form resulting from mechanical
+      transformation or translation of a Source form, including but
+      not limited to compiled object code, generated documentation,
+      and conversions to other media types.
+
+      "Work" shall mean the work of authorship, whether in Source or
+      Object form, made available under the License, as indicated by a
+      copyright notice that is included in or attached to the work
+      (an example is provided in the Appendix below).
+
+      "Derivative Works" shall mean any work, whether in Source or Object
+      form, that is based on (or derived from) the Work and for which the
+      editorial revisions, annotations, elaborations, or other modifications
+      represent, as a whole, an original work of authorship. For the purposes
+      of this License, Derivative Works shall not include works that remain
+      separable from, or merely link (or bind by name) to the interfaces of,
+      the Work and Derivative Works thereof.
+
+      "Contribution" shall mean any work of authorship, including
+      the original version of the Work and any modifications or additions
+      to that Work or Derivative Works thereof, that is intentionally
+      submitted to Licensor for inclusion in the Work by the copyright owner
+      or by an individual or Legal Entity authorized to submit on behalf of
+      the copyright owner. For the purposes of this definition, "submitted"
+      means any form of electronic, verbal, or written communication sent
+      to the Licensor or its representatives, including but not limited to
+      communication on electronic mailing lists, source code control systems,
+      and issue tracking systems that are managed by, or on behalf of, the
+      Licensor for the purpose of discussing and improving the Work, but
+      excluding communication that is conspicuously marked or otherwise
+      designated in writing by the copyright owner as "Not a Contribution."
+
+      "Contributor" shall mean Licensor and any individual or Legal Entity
+      on behalf of whom a Contribution has been received by Licensor and
+      subsequently incorporated within the Work.
+
+   2. Grant of Copyright License. Subject to the terms and conditions of
+      this License, each Contributor hereby grants to You a perpetual,
+      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
+      copyright license to reproduce, prepare Derivative Works of,
+      publicly display, publicly perform, sublicense, and distribute the
+      Work and such Derivative Works in Source or Object form.
+
+   3. Grant of Patent License. Subject to the terms and conditions of
+      this License, each Contributor hereby grants to You a perpetual,
+      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
+      (except as stated in this section) patent license to make, have made,
+      use, offer to sell, sell, import, and otherwise transfer the Work,
+      where such license applies only to those patent claims licensable
+      by such Contributor that are necessarily infringed by their
+      Contribution(s) alone or by combination of their Contribution(s)
+      with the Work to which such Contribution(s) was submitted. If You
+      institute patent litigation against any entity (including a
+      cross-claim or counterclaim in a lawsuit) alleging that the Work
+      or a Contribution incorporated within the Work constitutes direct
+      or contributory patent infringement, then any patent licenses
+      granted to You under this License for that Work shall terminate
+      as of the date such litigation is filed.
+
+   4. Redistribution. You may reproduce and distribute copies of the
+      Work or Derivative Works thereof in any medium, with or without
+      modifications, and in Source or Object form, provided that You
+      meet the following conditions:
+
+      (a) You must give any other recipients of the Work or
+          Derivative Works a copy of this License; and
+
+      (b) You must cause any modified files to carry prominent notices
+          stating that You changed the files; and
+
+      (c) You must retain, in the Source form of any Derivative Works
+          that You distribute, all copyright, patent, trademark, and
+          attribution notices from the Source form of the Work,
+          excluding those notices that do not pertain to any part of
+          the Derivative Works; and
+
+      (d) If the Work includes a "NOTICE" text file as part of its
+          distribution, then any Derivative Works that You distribute must
+          include a readable copy of the attribution notices contained
+          within such NOTICE file, excluding those notices that do not
+          pertain to any part of the Derivative Works, in at least one
+          of the following places: within a NOTICE text file distributed
+          as part of the Derivative Works; within the Source form or
+          documentation, if provided along with the Derivative Works; or,
+          within a display generated by the Derivative Works, if and
+          wherever such third-party notices normally appear. The contents
+          of the NOTICE file are for informational purposes only and
+          do not modify the License. You may add Your own attribution
+          notices within Derivative Works that You distribute, alongside
+          or as an addendum to the NOTICE text from the Work, provided
+          that such additional attribution notices cannot be construed
+          as modifying the License.
+
+      You may add Your own copyright statement to Your modifications and
+      may provide additional or different license terms and conditions
+      for use, reproduction, or distribution of Your modifications, or
+      for any such Derivative Works as a whole, provided Your use,
+      reproduction, and distribution of the Work otherwise complies with
+      the conditions stated in this License.
+
+   5. Submission of Contributions. Unless You explicitly state otherwise,
+      any Contribution intentionally submitted for inclusion in the Work
+      by You to the Licensor shall be under the terms and conditions of
+      this License, without any additional terms or conditions.
+      Notwithstanding the above, nothing herein shall supersede or modify
+      the terms of any separate license agreement you may have executed
+      with Licensor regarding such Contributions.
+
+   6. Trademarks. This License does not grant permission to use the trade
+      names, trademarks, service marks, or product names of the Licensor,
+      except as required for reasonable and customary use in describing the
+      origin of the Work and reproducing the content of the NOTICE file.
+
+   7. Disclaimer of Warranty. Unless required by applicable law or
+      agreed to in writing, Licensor provides the Work (and each
+      Contributor provides its Contributions) on an "AS IS" BASIS,
+      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+      implied, including, without limitation, any warranties or conditions
+      of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
+      PARTICULAR PURPOSE. You are solely responsible for determining the
+      appropriateness of using or redistributing the Work and assume any
+      risks associated with Your exercise of permissions under this License.
+
+   8. Limitation of Liability. In no event and under no legal theory,
+      whether in tort (including negligence), contract, or otherwise,
+      unless required by applicable law (such as deliberate and grossly
+      negligent acts) or agreed to in writing, shall any Contributor be
+      liable to You for damages, including any direct, indirect, special,
+      incidental, or consequential damages of any character arising as a
+      result of this License or out of the use or inability to use the
+      Work (including but not limited to damages for loss of goodwill,
+      work stoppage, computer failure or malfunction, or any and all
+      other commercial damages or losses), even if such Contributor
+      has been advised of the possibility of such damages.
+
+   9. Accepting Warranty or Additional Liability. While redistributing
+      the Work or Derivative Works thereof, You may choose to offer,
+      and charge a fee for, acceptance of support, warranty, indemnity,
+      or other liability obligations and/or rights consistent with this
+      License. However, in accepting such obligations, You may act only
+      on Your own behalf and on Your sole responsibility, not on behalf
+      of any other Contributor, and only if You agree to indemnify,
+      defend, and hold each Contributor harmless for any liability
+      incurred by, or claims asserted against, such Contributor by reason
+      of your accepting any such warranty or additional liability.
+
+   END OF TERMS AND CONDITIONS
+
+   APPENDIX: How to apply the Apache License to your work.
+
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "{}"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright {yyyy} {name of copyright owner}
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
diff --git a/playbooks/roles/repo_build/README.rst b/playbooks/roles/repo_build/README.rst
new file mode 100644
index 0000000..5252a47
--- /dev/null
+++ b/playbooks/roles/repo_build/README.rst
@@ -0,0 +1,14 @@
+OpenStack repo build
+#####################
+:tags: openstack, repo, build, cloud, ansible
+:category: \*nix
+
+Role to deploy a repository build for both python packages and git sources.
+
+.. code-block:: yaml
+
+    - name: Setup repo builds
+      hosts: repo_all
+      user: root
+      roles:
+        - { role: "repo_build", tags: [ "repo-build" ] }
diff --git a/playbooks/roles/repo_build/defaults/main.yml b/playbooks/roles/repo_build/defaults/main.yml
new file mode 100644
index 0000000..2a866c8
--- /dev/null
+++ b/playbooks/roles/repo_build/defaults/main.yml
@@ -0,0 +1,48 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This is disabled in Kilo due to the upper-constraints file not be fully backed in kilo
+repo_build_use_upper_constraints: false 
+
+repo_build_service_user_name: "nginx"
+
+repo_build_global_links_path: "/var/www/repo/links"
+repo_build_release_path: "/var/www/repo/os-releases"
+repo_build_dir: "/tmp/openstack-builder"
+repo_build_output: "/tmp/openstack-wheel-output"
+repo_build_git_dir: "/var/www/repo/openstackgit"
+repo_build_pool_dir: "/var/www/repo/pools"
+
+repo_build_release_tag: "untagged"
+
+repo_build_pip_default_index: "https://pypi.python.org/simple"
+repo_build_pip_extra_index: "https://pypi.python.org/simple"
+
+repo_build_timeout: 120
+
+repo_build_venv_force_rebuild: false
+repo_build_venv_build_dir: "/tmp/openstack-venv-builder"
+repo_build_venv_dir: "/var/www/repo/venvs"
+repo_build_venv_pip_install_options: >
+  --timeout 120
+  --find-links {{ repo_build_release_path }}/{{ repo_build_release_tag }}
+  --no-index
+  --verbose
+  --log /var/log/repo/repo_venv_builder.log
+
+# The venv building part of the repo infrastrucutre in kilo has been disabled
+#  because they're not used within the Kilo environment.
+repo_build_venv_create: false
+
diff --git a/playbooks/roles/repo_build/meta/main.yml b/playbooks/roles/repo_build/meta/main.yml
new file mode 100644
index 0000000..45f6aef
--- /dev/null
+++ b/playbooks/roles/repo_build/meta/main.yml
@@ -0,0 +1,33 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+galaxy_info:
+  author: rcbops
+  description: Install package repo build
+  company: Rackspace
+  license: Apache2
+  min_ansible_version: 1.9.3
+  platforms:
+    - name: Ubuntu
+      versions:
+        - trusty
+  categories:
+    - cloud
+    - python
+    - development
+    - openstack
+dependencies:
+  - apt_package_pinning
+  - pip_install
diff --git a/playbooks/roles/repo_build/tasks/main.yml b/playbooks/roles/repo_build/tasks/main.yml
new file mode 100644
index 0000000..05af212
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/main.yml
@@ -0,0 +1,27 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Wheel building
+- include: repo_clone_git.yml
+- include: repo_set_facts.yml
+- include: repo_pre_build.yml
+- include: repo_build.yml
+- include: repo_post_build.yml
+
+# Venv building
+- include: repo_venv.yml
+  when: repo_build_venv_create | bool
+  tags:
+    - repo-build-venvs
diff --git a/playbooks/roles/repo_build/tasks/repo_build.yml b/playbooks/roles/repo_build/tasks/repo_build.yml
new file mode 100644
index 0000000..f6230ce
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_build.yml
@@ -0,0 +1,50 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Create OpenStack global requirements wheels
+  shell: |
+    pip wheel --timeout {{ repo_build_timeout }} \
+              --wheel-dir {{ repo_build_output }} \
+              --allow-all-external \
+              --find-links {{ repo_build_global_links_path }} \
+              --constraint {{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_constraints.txt \
+              --index-url {{ repo_build_pip_default_index }} \
+              --trusted-host {{ repo_build_pip_default_index | netloc_no_port }} \
+              --extra-index-url {{ repo_build_pip_extra_index }} \
+              --trusted-host {{ repo_build_pip_extra_index | netloc_no_port }} \
+              --build {{ repo_build_dir }} \
+              --verbose \
+              --log /var/log/repo/repo_builder.log \
+              --requirement {{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_global_requirements.txt
+  tags:
+    - repo-build-global-requirement-wheels
+
+- name: Create OpenStack-Ansible requirement wheels
+  shell: |
+    pip wheel --timeout {{ repo_build_timeout }} \
+              --wheel-dir {{ repo_build_output }} \
+              --allow-all-external \
+              --find-links {{ repo_build_output }} \
+              --constraint {{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_constraints.txt \
+              --index-url {{ repo_build_pip_default_index }} \
+              --trusted-host {{ repo_build_pip_default_index | netloc_no_port }} \
+              --extra-index-url {{ repo_build_pip_extra_index }} \
+              --trusted-host {{ repo_build_pip_extra_index | netloc_no_port }} \
+              --build {{ repo_build_dir }} \
+              --verbose \
+              --log /var/log/repo/repo_builder.log \
+              --requirement {{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_local_filtered.txt
+  tags:
+    - repo-build-openstack-ansible-requirement-wheels
diff --git a/playbooks/roles/repo_build/tasks/repo_clone_git.yml b/playbooks/roles/repo_build/tasks/repo_clone_git.yml
new file mode 100644
index 0000000..eb5ab63
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_clone_git.yml
@@ -0,0 +1,25 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Clone all upstream git repositories
+  git:
+    repo: "{{ item['url'] }}"
+    dest: "{{ repo_build_git_dir }}/{{ item['name'] }}"
+    clone: yes
+    update: yes
+    version: "{{ item['version'] }}"
+  with_items: local_packages.results.0.item.remote_package_parts
+  tags:
+    - repo-clone-repos
diff --git a/playbooks/roles/repo_build/tasks/repo_post_build.yml b/playbooks/roles/repo_build/tasks/repo_post_build.yml
new file mode 100644
index 0000000..7dc2cce
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_post_build.yml
@@ -0,0 +1,116 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Index built wheels
+  command: "ls -1 {{ repo_build_output }}"
+  register: built_wheels
+  tags:
+    - repo-build-index-wheels
+    - repo-create-pool
+    - repo-copy-wheels-to-pool
+    - repo-create-release-links
+    - repo-create-links-index
+    - repo-create-release-manifest
+
+- name: Create wheel pool structure
+  file:
+    path: "{{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}"
+    state: directory
+    owner: "{{ repo_build_service_user_name }}"
+  with_items: built_wheels.stdout_lines
+  tags:
+    - repo-create-pool
+
+- name: Remove pool indexes if found
+  file:
+    path: "{{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/index.html"
+    state: absent
+  with_items: built_wheels.stdout_lines
+  tags:
+    - repo-create-pool
+
+- name: Move wheels into place and ensure permissions
+  shell: |
+    if [ ! -f "{{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/{{ item | lower }}" ];then
+      mv {{ repo_build_output }}/{{ item }} {{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/{{ item | lower }}
+    elif ! diff {{ repo_build_output }}/{{ item }} {{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/{{ item | lower }} > /dev/null;then
+      mv {{ repo_build_output }}/{{ item }} {{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/{{ item | lower }}
+    fi
+    chown {{ repo_build_service_user_name }} {{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/{{ item | lower }}
+  with_items: built_wheels.stdout_lines
+  tags:
+    - repo-copy-wheels-to-pool
+
+- name: Create release pool links
+  file:
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/{{ item | lower }}"
+    src: "{{ repo_build_pool_dir }}/{{ item.split('-')[0] | lower }}/{{ item | lower }}"
+    state: link
+    owner: "{{ repo_build_service_user_name }}"
+  with_items: built_wheels.stdout_lines
+  tags:
+    - repo-create-release-links
+
+- name: Create release manifest
+  template:
+    src: "manifest.in.j2"
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/MANIFEST.in"
+  tags:
+    - repo-create-release-manifest
+
+- name: Create absolute requirements
+  template:
+    src: "requirements_absolute_requirements.txt.j2"
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_absolute_requirements.txt"
+  tags:
+    - repo-create-absolute-requirements
+
+- name: Index built wheels
+  command: "ls -1 {{ repo_build_release_path }}/{{ repo_build_release_tag }}"
+  register: indexed_links
+  tags:
+    - repo-index-links
+    - repo-create-release-index
+
+- name: Create release index
+  template:
+    src: "release_index_links.html.j2"
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/index.html"
+  tags:
+    - repo-create-release-index
+
+- name: Create general index links
+  file:
+    dest: "{{ repo_build_global_links_path }}/{{ item | lower }}"
+    src: "../pools/{{ item.split('-')[0] | lower }}/{{ item | lower }}"
+    state: link
+    owner: "{{ repo_build_service_user_name }}"
+  with_items: built_wheels.stdout_lines
+  tags:
+    - repo-create-links-index
+
+- name: Index general links
+  command: "ls -1 {{ repo_build_global_links_path }}"
+  register: global_indexed_links
+  tags:
+    - repo-index-links
+    - repo-create-global-release-index
+
+- name: Create release index
+  template:
+    src: "global_indexed_links.html.j2"
+    dest: "{{ repo_build_global_links_path }}/index.html"
+  tags:
+    - repo-create-global-release-index
diff --git a/playbooks/roles/repo_build/tasks/repo_pre_build.yml b/playbooks/roles/repo_build/tasks/repo_pre_build.yml
new file mode 100644
index 0000000..4f3ccfe
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_pre_build.yml
@@ -0,0 +1,58 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Ensure workspace files are cleaned up
+  file:
+    path: "{{ item }}"
+    state: "absent"
+  with_items:
+    - "{{ repo_build_dir }}"
+    - "{{ repo_build_output }}"
+    - "{{ repo_build_release_path }}/{{ repo_build_release_tag }}"
+    - "{{ repo_build_pool_dir }}/index.html"  # pool directory index is removed because its no longer used
+  tags:
+    - repo-clean-workspace
+
+- name: Create release directory
+  file:
+    path: "{{ item }}"
+    state: directory
+    owner: "{{ repo_build_service_user_name }}"
+  with_items:
+    - "{{ repo_build_release_path }}/{{ repo_build_release_tag }}"
+    - "{{ repo_build_global_links_path }}"
+  tags:
+    - repo-create-release-links-location
+
+- name: Build global package requirements file
+  template:
+    src: "requirements_global_requirements.txt.j2"
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_global_requirements.txt"
+  tags:
+    - repo-build-global-package-files
+
+- name: Build filtered package requirements file
+  template:
+    src: "requirements_local_filtered.txt.j2"
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_local_filtered.txt"
+  tags:
+    - repo-build-filtered-package-files
+
+- name: Build package constraints file
+  template:
+    src: "requirements_constraints.txt.j2"
+    dest: "{{ repo_build_release_path }}/{{ repo_build_release_tag }}/requirements_constraints.txt"
+  tags:
+    - repo-build-constraints-file
diff --git a/playbooks/roles/repo_build/tasks/repo_set_facts.yml b/playbooks/roles/repo_build/tasks/repo_set_facts.yml
new file mode 100644
index 0000000..fc7e5cd
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_set_facts.yml
@@ -0,0 +1,64 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Retrieve global requirements content
+  slurp:
+    src: "{{ repo_build_git_dir }}/requirements/global-requirements.txt"
+  register: slurp_global_requirements
+  tags:
+    - repo-set-requirement-names-filtered
+    - repo-set-requirement-names
+    - repo-set-requirements
+    - repo-get-global-requirements
+
+- name: Set requirements
+  set_fact:
+    global_requirement: "{{ slurp_global_requirements.content | b64decode | splitlines }}"
+  tags:
+    - repo-set-requirement-names-filtered
+    - repo-set-requirement-names
+    - repo-set-requirements
+
+- name: Set requirement names
+  set_fact:
+    global_requirement_names: "{{ global_requirement | pip_requirement_names }}"
+  tags:
+    - repo-set-requirement-names-filtered
+    - repo-set-requirement-names
+    - repo-set-requirements
+
+- name: Set filtered requirement names
+  set_fact:
+    global_requirement_names_filtered: "{{ local_packages.results.0.item.packages | filtered_list(global_requirement_names) }}"
+  tags:
+    - repo-set-requirement-names-filtered
+    - repo-set-requirements
+
+- name: Retrieve upper constraints content
+  slurp:
+    src: "{{ repo_build_git_dir }}/requirements/upper-constraints.txt"
+  register: slurp_upper_constraints
+  tags:
+    - repo-set-constraints
+    - repo-get-upper-constraints
+    - repo-build-constraints-file
+
+- name: Set upper constraints
+  set_fact:
+    upper_constraints: "{{ slurp_upper_constraints.content | b64decode | splitlines }}"
+  when: slurp_upper_constraints | success
+  tags:
+    - repo-set-constraints
+    - repo-build-constraints-file
diff --git a/playbooks/roles/repo_build/tasks/repo_venv.yml b/playbooks/roles/repo_build/tasks/repo_venv.yml
new file mode 100644
index 0000000..12cd64b
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_venv.yml
@@ -0,0 +1,18 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- include: repo_venv_pre_build.yml
+- include: repo_venv_build.yml
+- include: repo_venv_post_build.yml
diff --git a/playbooks/roles/repo_build/tasks/repo_venv_build.yml b/playbooks/roles/repo_build/tasks/repo_venv_build.yml
new file mode 100644
index 0000000..a97fc01
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_venv_build.yml
@@ -0,0 +1,81 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Get venv command path
+  command: which virtualenv
+  register: virtualenv_path
+  tags:
+    - repo-command-bin
+    - repo-create-venv
+
+- name: Set virtualenv command path
+  set_fact:
+    virtualenv_bin: "{{ virtualenv_path.stdout }}"
+  tags:
+    - repo-command-bin
+    - repo-create-venv
+
+- name: Check for created venvs
+  command: >
+    ls -1 "{{ repo_build_venv_dir }}/{{ repo_build_release_tag }}/{{ ansible_distribution | lower }}/"
+  register: created_venvs
+  tags:
+    - repo-create-venv
+
+- name: Set existing venv fact
+  set_fact:
+    existing_venvs: "{{ created_venvs.stdout_lines }}"
+  tags:
+    - repo-command-bin
+    - repo-create-venv
+
+- name: Create role based venv
+  pip:
+    name: "{{ item.value | join(' ') }}"
+    state: present
+    virtualenv: "{{ repo_build_venv_build_dir }}/venvs/{{ item.key | replace('os_', '') }}"
+    virtualenv_site_packages: "no"
+    virtualenv_command: "{{ virtualenv_bin }} --always-copy"
+    extra_args: "{{ repo_build_venv_pip_install_options }}"
+  register: install_packages
+  until: install_packages|success
+  retries: 5
+  delay: 2
+  with_dict: local_packages.results.0.item.role_packages
+  when:
+    - '"os_" in item.key'
+    - "'{{ item.key | replace('os_', '') }}-{{ repo_build_release_tag }}.tgz' not in existing_venvs"
+  tags:
+    - repo-create-venv
+
+- name: Create venv archive
+  shell: |
+    # This is to clean up pyc files which makes the archived venv smaller
+    #  TODO(cloudnull) This should use the find module when we move to Ansible 2.0
+    find "{{ repo_build_venv_build_dir }}/venvs/{{ item.key | replace('os_', '') }}" -name '*.pyc' -delete
+    # Create archive
+    tar czf "{{ item.key | replace('os_', '') }}-{{ repo_build_release_tag }}.tgz" \
+        -C "{{ repo_build_venv_build_dir }}/venvs/{{ item.key | replace('os_', '') }}" .
+  args:
+    chdir: "{{ repo_build_venv_dir }}/{{ repo_build_release_tag }}/{{ ansible_distribution | lower }}"
+    creates: "{{ repo_build_venv_dir }}/{{ repo_build_release_tag }}/{{ ansible_distribution | lower }}/{{ item.key | replace('os_', '') }}-{{ repo_build_release_tag }}.tgz"
+  with_dict: local_packages.results.0.item.role_packages
+  when:
+    - '"os_" in item.key'
+  tags:
+    - skip_ansible_lint
+    - repo-venv-compress-archive
+    - repo-create-venv-archive
+    - repo-create-venv
diff --git a/playbooks/roles/repo_build/tasks/repo_venv_post_build.yml b/playbooks/roles/repo_build/tasks/repo_venv_post_build.yml
new file mode 100644
index 0000000..a28ca5b
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_venv_post_build.yml
@@ -0,0 +1,23 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Cleanup venv directory
+  file:
+    path: "{{ item }}"
+    state: absent
+  with_items:
+    - "{{ repo_build_venv_build_dir }}"
+  tags:
+    - repo-cleanup-venv-location
diff --git a/playbooks/roles/repo_build/tasks/repo_venv_pre_build.yml b/playbooks/roles/repo_build/tasks/repo_venv_pre_build.yml
new file mode 100644
index 0000000..da8b21f
--- /dev/null
+++ b/playbooks/roles/repo_build/tasks/repo_venv_pre_build.yml
@@ -0,0 +1,49 @@
+---
+# Copyright 2015, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Make sure old venv build directories are clean
+  file:
+    path: "{{ item }}"
+    state: "absent"
+  with_items:
+    - "{{ repo_build_venv_build_dir }}"
+  tags:
+    - repo-create-venv-location
+    - repo-venv-compress-archive
+    - repo-create-venv-archive
+
+- name: Destroy base venvs to rebuild them
+  file:
+    path: "{{ repo_build_venv_dir }}/{{ repo_build_release_tag }}/{{ ansible_distribution | lower }}"
+    state: "absent"
+  when: repo_build_venv_force_rebuild | bool
+  tags:
+    - repo-create-venv-location
+    - repo-venv-compress-archive
+    - repo-create-venv-archive
+
+- name: Create venv directory
+  file:
+    path: "{{ item }}"
+    state: "directory"
+    owner: "{{ repo_build_service_user_name }}"
+    mode: "2755"
+  with_items:
+    - "{{ repo_build_venv_build_dir }}/venvs"
+    - "{{ repo_build_venv_dir }}/{{ repo_build_release_tag }}/{{ ansible_distribution | lower }}"
+  tags:
+    - repo-create-venv-location
+    - repo-venv-compress-archive
+    - repo-create-venv-archive
diff --git a/playbooks/roles/repo_build/templates/global_indexed_links.html.j2 b/playbooks/roles/repo_build/templates/global_indexed_links.html.j2
new file mode 100644
index 0000000..efb0bd6
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/global_indexed_links.html.j2
@@ -0,0 +1,12 @@
+<html>
+<head>
+<title>
+links for "{{ repo_build_release_tag }}"
+</title>
+</head>
+<body>
+{% for item in global_indexed_links.stdout_lines %}
+<a href="{{ item }}">{{ item }}</a><br />
+{% endfor %}
+</body>
+</html>
diff --git a/playbooks/roles/repo_build/templates/manifest.in.j2 b/playbooks/roles/repo_build/templates/manifest.in.j2
new file mode 100644
index 0000000..7c6b5ae
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/manifest.in.j2
@@ -0,0 +1,7 @@
+{% for item in built_wheels.stdout_lines %}
+{{ repo_build_pool_dir | basename }}/{{ item.split('-')[0] | lower }}/{{ item | lower }}
+{{ repo_build_release_path | basename }}/{{ repo_build_release_tag }}/{{ item | lower }}
+{% endfor %}
+{% for clone_item in local_packages.results.0.item.remote_package_parts -%}
+{{ repo_build_git_dir | basename }}/{{ clone_item['name'] }}
+{% endfor %}
\ No newline at end of file
diff --git a/playbooks/roles/repo_build/templates/release_index_links.html.j2 b/playbooks/roles/repo_build/templates/release_index_links.html.j2
new file mode 100644
index 0000000..36b7fc8
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/release_index_links.html.j2
@@ -0,0 +1,12 @@
+<html>
+<head>
+<title>
+links for "{{ repo_build_release_tag }}"
+</title>
+</head>
+<body>
+{% for item in indexed_links.stdout_lines %}
+<a href="{{ item }}">{{ item }}</a><br />
+{% endfor %}
+</body>
+</html>
\ No newline at end of file
diff --git a/playbooks/roles/repo_build/templates/requirements_absolute_requirements.txt.j2 b/playbooks/roles/repo_build/templates/requirements_absolute_requirements.txt.j2
new file mode 100644
index 0000000..9183186
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/requirements_absolute_requirements.txt.j2
@@ -0,0 +1,3 @@
+{% for item in built_wheels.stdout_lines %}
+{{ item.split('-')[0] | lower }}=={{ item.split('-')[1] | lower }}
+{% endfor %}
\ No newline at end of file
diff --git a/playbooks/roles/repo_build/templates/requirements_constraints.txt.j2 b/playbooks/roles/repo_build/templates/requirements_constraints.txt.j2
new file mode 100644
index 0000000..11a75c6
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/requirements_constraints.txt.j2
@@ -0,0 +1,21 @@
+# Computed constraints
+{% set constraint_pkgs = [] -%}
+{% for clone_item in local_packages.results.0.item.remote_package_parts -%}
+{% if 'ignorerequirements=true' not in clone_item['original'] %}
+git+file://{{ repo_build_git_dir }}/{{ clone_item['name'] }}@{{ clone_item['version'] }}#egg={{ clone_item['name'] | replace('-', '_') | lower }}
+{% set _ = constraint_pkgs.append(clone_item['name'] | replace('-', '_') | lower) %}
+{% endif %}
+{% endfor %}
+# upper boundry constraints from requirements repo.
+{% for constraint_item in upper_constraints %}
+{%- set constraint_split = constraint_item.split('===') %}
+{%- set constraint_name = constraint_split[0] %}
+{%- set constraint_name_normalized = constraint_name | replace('-', '_') | lower %}
+{% if constraint_name_normalized not in constraint_pkgs %}
+{% if repo_build_use_upper_constraints | bool %}
+{{ constraint_split[0] | replace('-', '_') | lower }}<={{ constraint_split[1] }}
+{% else %}
+# {{ constraint_split[0] | replace('-', '_') | lower }}<={{ constraint_split[1] }}
+{% endif %}
+{% endif %}
+{% endfor %}
diff --git a/playbooks/roles/repo_build/templates/requirements_global_requirements.txt.j2 b/playbooks/roles/repo_build/templates/requirements_global_requirements.txt.j2
new file mode 100644
index 0000000..c11cfd4
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/requirements_global_requirements.txt.j2
@@ -0,0 +1,3 @@
+{% for item in global_requirement %}
+{{ item }}
+{% endfor %}
\ No newline at end of file
diff --git a/playbooks/roles/repo_build/templates/requirements_local_filtered.txt.j2 b/playbooks/roles/repo_build/templates/requirements_local_filtered.txt.j2
new file mode 100644
index 0000000..846c539
--- /dev/null
+++ b/playbooks/roles/repo_build/templates/requirements_local_filtered.txt.j2
@@ -0,0 +1,3 @@
+{% for item in global_requirement_names_filtered %}
+{{ item }}
+{% endfor %}
\ No newline at end of file
diff --git a/playbooks/roles/repo_server/defaults/main.yml b/playbooks/roles/repo_server/defaults/main.yml
index 963d61b..d2cb481 100644
--- a/playbooks/roles/repo_server/defaults/main.yml
+++ b/playbooks/roles/repo_server/defaults/main.yml
@@ -85,6 +85,8 @@
   - turbolift
   - wheel
   - yaprt
+  - virtualenv
+  - virtualenv-tools
 
 # Main web server port
 repo_server_port: 8181
